library/data-model/src/permission/model.ts

```diff
@@ -706,9 +889,64 @@ export const roleActions: Record<
       Action.INITIALISE_SYSTEM_API,
       Action.RESTORE_FROM_BACKUP,
       Action.VALIDATE_DBS,
+      Action.CREATE_TEAM,
       Action.SEND_TEST_EMAIL,
     ],
-    inheritedRoles: [Role.GENERAL_CREATOR, Role.PROJECT_ADMIN],
+    inheritedRoles: [
+      // God role
+      Role.GENERAL_CREATOR,
+      Role.PROJECT_ADMIN,
+      Role.TEAM_ADMIN,
+      Role.TEMPLATE_ADMIN,
+    ],
+  },
+
+  // TEAM ROLES
+  [Role.TEAM_MEMBER]: {
+    actions: [Action.VIEW_TEAM_DETAILS, Action.VIEW_TEAM_MEMBERS],
+    virtualRoles: new Map([
+      // Projects owned by team -> contributor
+      [Resource.PROJECT, [Role.PROJECT_CONTRIBUTOR]],
+      // Template owned by team -> guest
+      [Resource.TEMPLATE, [Role.TEMPLATE_GUEST]],
+    ]),
+  },
+
+  [Role.TEAM_MANAGER]: {
+    actions: [
+      // Other team based things
+      Action.UPDATE_TEAM_DETAILS, // resource
+      Action.ADD_MEMBER_TO_TEAM, // resource
+      Action.REMOVE_MEMBER_FROM_TEAM, // resource
+
+      // Being a manager of a team also gives you the ability to create new key
+      // resources in that team
+      Action.CREATE_PROJECT_IN_TEAM, // resource
+      Action.CREATE_TEMPLATE_IN_TEAM, // resource
+    ],
+    // TODO make templates managed with their own roles/ownership model so that
+    // we can do virtualRoles over templates owned by the team
+    // NOTE this is a bit of a permission leak here re: general creator
+    inheritedRoles: [Role.TEAM_MEMBER],
+    // Projects owned by team -> manager
+    virtualRoles: new Map([[Resource.PROJECT, [Role.PROJECT_MANAGER]]]),
+  },
+
+  [Role.TEAM_ADMIN]: {
+    actions: [
+      Action.DELETE_TEAM,
+      Action.ADD_ADMIN_TO_TEAM,
+      Action.REMOVE_ADMIN_FROM_TEAM,
+      Action.ADD_MANAGER_TO_TEAM,
+      Action.REMOVE_MANAGER_FROM_TEAM,
+    ],
+    inheritedRoles: [Role.TEAM_MANAGER],
+    virtualRoles: new Map([
+      // Projects owned by team -> manager
+      [Resource.PROJECT, [Role.PROJECT_ADMIN]],
+      // Template owned by team -> admin
+      [Resource.TEMPLATE, [Role.TEMPLATE_ADMIN]],
+    ]),
```

Comment:
This module is quite long now, I wonder if we might want to refactor it to more manageable modules at some point?

Reply:
Yeah - I would agree - could split into actions, roles, model ? Let's do later